xcpmd/src/battery.c

```diff
@@ -0,0 +1,803 @@
+int update_battery_status(unsigned int battery_index) {
+
+    int i, rc;
+    DIR *battery_dir;
+    struct dirent * dp;
+    FILE *file;
+    char filename[256];
+    char data[128];
+    char *ptr;
+
+    struct battery_status status;
+    memset(&status, 0, sizeof(struct battery_status));
+
+    if (battery_index >= num_battery_structs_allocd)
+        return -1;
+
+    if (battery_slot_exists(battery_index) == NO) {
+        status.present = NO;
+        memcpy(&last_status[battery_index], &status, sizeof(struct battery_status));
+        return 1;
+    }
+
+    battery_dir = get_battery_dir(battery_index);
+    if (!battery_dir) {
+        //Battery directory does not exist--this normally occurs when a battery slot is removed
+        if (errno == ENOENT) {
+            status.present = NO;
+            //memcpy(&last_status[battery_index], &status, sizeof(struct battery_status));
+            return 1;
+        }
+        else {
+            xcpmd_log(LOG_ERR, "opendir in update_battery_status for directory %s/BAT%d failed with error %d\n", BATTERY_DIR_PATH, battery_index, errno);
+            return 0;
+        }
+    }
+
+    //Loop over the files in the directory.
+    while ((dp = readdir(battery_dir)) != NULL) {
+
+        //Convert from dirent to file and read out the data.
+        if (dp->d_type == DT_REG) {
+
+            memset(filename, 0, sizeof(filename));
+            sprintf(filename, "%s/BAT%i/%s", BATTERY_DIR_PATH, battery_index, dp->d_name);
+
+            file = fopen(filename, "r");
+            if (file == NULL)
+                continue;
+
+            memset(data, 0, sizeof(data));
+            fgets(data, sizeof(data), file);
+
+            fclose(file);
+
+            //Trim off leading spaces.
+            ptr = data;
+            while(*ptr == ' ')
+                ptr += sizeof(char);
+
+            //Set the attribute represented by this file.
+            set_battery_status_attribute(dp->d_name, ptr, &status);
+        }
+    }
+
+    // This check handles both cases for mA batteries: if are not charging
+    // (current_now == 0) but have capacity or the battery is totally dead
+    // (charge_now == 0) but it is charging. If both are zero, they will
+    // both be zero in the end.
+    if (status.charge_now != 0 || status.current_now != 0) {
+        // Rate in mA, remaining in mAh
+        status.present_rate = status.current_now;
+        status.remaining_capacity = status.charge_now;
+    }
+    else {
+        // Rate in mW, remaining in mWh
+        status.present_rate = status.power_now;
+        status.remaining_capacity = status.energy_now;
+    }
+
+    closedir(battery_dir);
+    memcpy(&last_status[battery_index], &status, sizeof(struct battery_status));
+    //print_battery_status();
+    return 1;
+}
+
+
+//Gets a battery's info from the sysfs and stores it in last_info.
+int update_battery_info(unsigned int battery_index) {
```

Comment:
This is a leftover from the event-driven code. With the current architecture, it would definitely be more efficient to merge this function with the previous one.